# Wrangle Pisa 2012
#### By Gabriela Sikora


### Abstract



### Table of Contents

<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#wrangling">Data Wrangling</a></li>
    <ul>
    <li><a href="#gathering">Gathering Data</a></li>
    <li><a href="#assessing1">Initial Assessment</a></li>
        <ul>
          <li><a href="#issues1">Identified Issues</a></li>
        </ul>
    <li><a href="#cleaning1">Cleaning Data</a></li>
    <li><a href="#assessing2">Final Assessment</a></li>
    </ul>
<li><a href="#sav">Store</a></li>
<li><a href="#conclusion">Conclusion</a></li>
<li><a href="#references">References</a></li>    
</ul>

<a id='intro'></a>
## Introduction



To begin, we first need to import all the relevant libraries:
#### Import

In [1]:
# import all packages
import numpy as np
import pandas as pd

<a id='wrangling'></a>
## Data Wrangling

In this section of the report, we will load in the data, check it for cleanliness, and then trim and clean the datasets for analysis. 

<a id='gathering'></a>
### Gathering Data

Unfortunately the Pisa dataset is too large to be uploaded to github, but if you would like to work with it yourself then place follow this link: https://s3.amazonaws.com/udacity-hosted-downloads/ud507/pisa2012.csv.zip 

In [2]:
# Encoding help found here: https://stackoverflow.com/questions/30462807/encoding-error-in-panda-read-csv

# Load in the Pisa dataframe
pisa = pd.read_csv('pisa2012.csv', encoding="cp1252", sep=',')

# Load in the Pisa dictionary
pisa_dict = pd.read_csv('pisadict2012.csv', encoding="cp1252", sep=',')

/Users/gabriela/Desktop/Code/myenv/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3044: DtypeWarning: Columns (15,16,17,21,22,23,24,25,26,30,31,36,37,45,65,123,155,156,157,158,159,160,161,162,163,164,165,166,167,168,169,170,171,284,285,286,287,288,289,290,291,292,293,294,295,296,297,298,299,300,301,302,303,307,308,309,310,311,312,313,314,315,316,317,318,319,320,321,322,323,324,325,326,327,328,329,330,331,332,333,334,335,336,337,338,339,340,341,342,343,344,345,346,347,348,349,350,351,352,353,354,355,356,357,376,377,378,379,380,381,382,383,384,385,386,387,388,389,390,391,392,393,394,395,396,397,398,399,400,401,402,403,475) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


<a id='assessing1'></a>
### First Assessment

In this section, we will do the preliminary visual and programmatic assessment of the pisa datasets to determine whether or not it holds any major quality or tidiness issues.

In [3]:
pisa

Unnamed: 0      CNT  SUBNATIO  STRATUM      OECD         NC  SCHOOLID  \
0                1  Albania     80000  ALB0006  Non-OECD    Albania         1   
1                2  Albania     80000  ALB0006  Non-OECD    Albania         1   
2                3  Albania     80000  ALB0006  Non-OECD    Albania         1   
3                4  Albania     80000  ALB0006  Non-OECD    Albania         1   
4                5  Albania     80000  ALB0006  Non-OECD    Albania         1   
5                6  Albania     80000  ALB0006  Non-OECD    Albania         1   
6                7  Albania     80000  ALB0006  Non-OECD    Albania         1   
7                8  Albania     80000  ALB0006  Non-OECD    Albania         1   
8                9  Albania     80000  ALB0006  Non-OECD    Albania         1   
9               10  Albania     80000  ALB0005  Non-OECD    Albania         2   
10              11  Albania     80000  ALB0005  Non-OECD    Albania         2   
11              12  Albania     80000  ALB0005  Non-OECD    Albania         2   
12              13  Albania     80000  ALB0005  Non-OECD    Albania         2   
13              14  Albania     80000  ALB0005  Non-OECD    Albania         2   
14              15  Albania     80000  ALB0005  Non-OECD    Albania         2   
15              16  Albania     80000  ALB0005  Non-OECD    Albania         2   
16              17  Albania     80000  ALB0005  Non-OECD    Albania         2   
17              18  Albania     80000  ALB0005  Non-OECD    Albania         2   
18              19  Albania     80000  ALB0005  Non-OECD    Albania         2   
19              20  Albania     80000  ALB0005  Non-OECD    Albania         2   
20              21  Albania     80000  ALB0005  Non-OECD    Albania         2   
21              22  Albania     80000  ALB0005  Non-OECD    Albania         2   
22              23  Albania     80000  ALB0005  Non-OECD    Albania         2   
23              24  Albania     80000  ALB0005  Non-OECD    Albania         2   
24              25  Albania     80000  ALB0005  Non-OECD    Albania         2   
25              26  Albania     80000  ALB0005  Non-OECD    Albania         2   
26              27  Albania     80000  ALB0005  Non-OECD    Albania         2   
27              28  Albania     80000  ALB0005  Non-OECD    Albania         2   
28              29  Albania     80000  ALB0005  Non-OECD    Albania         2   
29              30  Albania     80000  ALB0005  Non-OECD    Albania         2   
...            ...      ...       ...      ...       ...        ...       ...   
485460      485461  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485461      485462  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485462      485463  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485463      485464  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485464      485465  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485465      485466  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485466      485467  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485467      485468  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485468      485469  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485469      485470  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485470      485471  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485471      485472  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485472      485473  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485473      485474  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485474      485475  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485475      485476  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485476      485477  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485477      485478  Vietnam   7040000  

In [4]:
pisa.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 485490 entries, 0 to 485489
Columns: 636 entries, Unnamed: 0 to VER_STU
dtypes: float64(250), int64(18), object(368)
memory usage: 2.3+ GB


In [5]:
pisa_dict

Unnamed: 0                                                  x
0           CNT                           Country code 3-character
1      SUBNATIO  Adjudicated sub-region code 7-digit code (3-di...
2       STRATUM  Stratum ID 7-character (cnt + region ID + orig...
3          OECD                                       OECD country
4            NC                       National Centre 6-digit Code
5      SCHOOLID  School ID 7-digit (region ID + stratum ID + 3-...
6       STIDSTD                                         Student ID
7       ST01Q01                                International Grade
8       ST02Q01                           National Study Programme
9       ST03Q01                                      Birth - Month
10      ST03Q02                                        Birth -Year
11      ST04Q01                                             Gender
12      ST05Q01                                   Attend <ISCED 0>
13      ST06Q01                                   Age at <ISCED 1>
14      ST07Q01                                 Repeat - <ISCED 1>
15      ST07Q02                                 Repeat - <ISCED 2>
16      ST07Q03                                 Repeat - <ISCED 3>
17      ST08Q01                          Truancy - Late for School
18      ST09Q01                    Truancy - Skip whole school day
19     ST115Q01           Truancy - Skip classes within school day
20      ST11Q01                                   At Home - Mother
21      ST11Q02                                   At Home - Father
22      ST11Q03                                 At Home - Brothers
23      ST11Q04                                  At Home - Sisters
24      ST11Q05                             At Home - Grandparents
25      ST11Q06                                   At Home - Others
26      ST13Q01                          Mother<Highest Schooling>
27      ST14Q01            Mother Qualifications - <ISCED level 6>
28      ST14Q02           Mother Qualifications - <ISCED level 5A>
29      ST14Q03           Mother Qualifications - <ISCED level 5B>
..          ...                                                ...
605    W_FSTR55           FINAL STUDENT REPLICATE BRR-FAY WEIGHT55
606    W_FSTR56           FINAL STUDENT REPLICATE BRR-FAY WEIGHT56
607    W_FSTR57           FINAL STUDENT REPLICATE BRR-FAY WEIGHT57
608    W_FSTR58           FINAL STUDENT REPLICATE BRR-FAY WEIGHT58
609    W_FSTR59           FINAL STUDENT REPLICATE BRR-FAY WEIGHT59
610    W_FSTR60           FINAL STUDENT REPLICATE BRR-FAY WEIGHT60
611    W_FSTR61           FINAL STUDENT REPLICATE BRR-FAY WEIGHT61
612    W_FSTR62           FINAL STUDENT REPLICATE BRR-FAY WEIGHT62
613    W_FSTR63           FINAL STUDENT REPLICATE BRR-FAY WEIGHT63
614    W_FSTR64           FINAL STUDENT REPLICATE BRR-FAY WEIGHT64
615    W_FSTR65           FINAL STUDENT REPLICATE BRR-FAY WEIGHT65
616    W_FSTR66           FINAL STUDENT REPLICATE BRR-FAY WEIGHT66
617    W_FSTR67           FINAL STUDENT REPLICATE BRR-FAY WEIGHT67
618    W_FSTR68           FINAL STUDENT REPLICATE BRR-FAY WEIGHT68
619    W_FSTR69           FINAL STUDENT REPLICATE BRR-FAY WEIGHT69
620    W_FSTR70           FINAL STUDENT REPLICATE BRR-FAY WEIGHT70
621    W_FSTR71           FINAL STUDENT REPLICATE BRR-FAY WEIGHT71
622    W_FSTR72           FINAL STUDENT REPLICATE BRR-FAY WEIGHT72
623    W_FSTR73           FINAL STUDENT REPLICATE BRR-FAY WEIGHT73
624    W_FSTR74           FINAL STUDENT REPLICATE BRR-FAY WEIGHT74
625    W_FSTR75           FINAL STUDENT REPLICATE BRR-FAY WEIGHT75
626    W_FSTR76           FINAL STUDENT REPLICATE BRR-FAY WEIGHT76
627    W_FSTR77           FINAL STUDENT REPLICATE BRR-FAY WEIGHT77
628    W_FSTR78           FINAL STUDENT REPLICATE BRR-FAY WEIGHT78
629    W_FSTR79           FINAL STUDENT REPLICATE BRR-FAY WEIGHT79
630    W_FSTR80           FINAL STUDENT REPLICATE BRR-FAY WEIGHT80
631    WVARSTRR           RANDOMIZED FINAL VARIANCE STRATUM (1-80)
632    VAR_UNIT                    RANDOMLY ASSIGNED

In [6]:
# Check for null values in relevant columns
columns = ['STIDSTD', 'ST01Q01', 'ST04Q01', 'ST13Q01', 
            'ST17Q01', 'ST28Q01', 'ST57Q01', 'ST57Q02', 'ST57Q03', 
            'ST57Q04', 'ST57Q05', 'MMINS', 'LMINS', 'SMINS',
            'PV1MATH', 'PV2MATH', 'PV3MATH', 'PV4MATH', 'PV5MATH',  
            'PV1READ', 'PV2READ',  'PV3READ', 'PV4READ', 'PV5READ',
            'PV1SCIE', 'PV2SCIE', 'PV3SCIE', 'PV4SCIE', 'PV5SCIE']

for x in columns:
    print(x, pisa[x].isna().sum())

STIDSTD 0
ST01Q01 0
ST04Q01 0
ST13Q01 27511
ST17Q01 42229
ST28Q01 11725
ST57Q01 184123
ST57Q02 215682
ST57Q03 201677
ST57Q04 205833
ST57Q05 195988
MMINS 202187
LMINS 202624
SMINS 214576
PV1MATH 0
PV2MATH 0
PV3MATH 0
PV4MATH 0
PV5MATH 0
PV1READ 0
PV2READ 0
PV3READ 0
PV4READ 0
PV5READ 0
PV1SCIE 0
PV2SCIE 0
PV3SCIE 0
PV4SCIE 0
PV5SCIE 0


In [7]:
pisa['CNT'].value_counts()

Mexico                      33806
Italy                       31073
Spain                       25313
Canada                      21544
Brazil                      19204
Australia                   14481
United Kingdom              12659
United Arab Emirates        11500
Switzerland                 11229
Qatar                       10966
Colombia                     9073
Finland                      8829
Belgium                      8597
Denmark                      7481
Jordan                       7038
Chile                        6856
Thailand                     6606
Japan                        6351
Chinese Taipei               6046
Peru                         6035
Slovenia                     5911
Argentina                    5908
Kazakhstan                   5808
Portugal                     5722
Indonesia                    5622
Singapore                    5546
Macao-China                  5335
Czech Republic               5327
Uruguay                      5315
Bulgaria      

In [8]:
pisa.duplicated().sum()

0

<a id='issues1'></a>
### Identified Issues
Here we can see some of the issues noticed in the dataframes. These quality issues will be cleaned in the following section.

1. Narrow down the dataframe by only looking at variables of interest since there are more variables than needed
2. Make the variable names readable since they are codes that can only be read with the dictionary
3. Create an average of all Math, Reading, and Science scores since all 5 scores for each are not necessary
4. Missing values in relevant columns ('Mother Highest Schooling', 'Father Highest Schooling', 'How many books at home', 'Out-of-School Study Time - Homework', 'Out-of-School Study Time - Guided Homework', 'Out-of-School Study Time - Personal Tutor','Out-of-School Study Time - Commercial Company', 'Out-of-School Study Time - With Parent', 'Learning time (minutes per week)- Mathematics', 'Learning time (minutes per week)  - test language', 'Learning time (minutes per week) - Science')
5. Create a total for Out-of-School Study Time and Learning time (minutes per week)
6. Single countries have various entries in the country column (Florida (USA), Massachusetts (USA), Connecticut (USA) and United States of America)


<a id='cleaning1'></a>
### First Cleaning

In [9]:
# Create a copy to preserve the original dataframe
pisa_clean = pisa.copy()

### Narrow down the dataframe by only looking at variables of interest

There are 636 columns available in the original dataframe, which is far more than what is necessary. We will rather look at just the following variables.

In [10]:
pisa_clean = pisa_clean[['STIDSTD', 'ST04Q01', 'ST13Q01', 'ST14Q01', 'ST14Q02', 'ST14Q03', 'ST14Q04', 
                         'ST17Q01', 'ST18Q01', 'ST18Q02', 'ST18Q03', 'ST18Q04', 
                         'ST28Q01', 
                         'ST57Q01', 'ST57Q02', 'ST57Q03', 'ST57Q04', 'ST57Q05', 'MMINS', 
                         'LMINS', 'SMINS', 'PV1MATH', 'PV2MATH', 'PV3MATH', 'PV4MATH', 
                         'PV5MATH', 'PV1READ', 'PV2READ',  'PV3READ', 'PV4READ', 'PV5READ',
                         'PV1SCIE', 'PV2SCIE', 'PV3SCIE', 'PV4SCIE', 'PV5SCIE']]       

In [11]:
# Check to see if the columns are correctly reduced
list(pisa_clean)

['STIDSTD',
 'ST04Q01',
 'ST13Q01',
 'ST14Q01',
 'ST14Q02',
 'ST14Q03',
 'ST14Q04',
 'ST17Q01',
 'ST18Q01',
 'ST18Q02',
 'ST18Q03',
 'ST18Q04',
 'ST28Q01',
 'ST57Q01',
 'ST57Q02',
 'ST57Q03',
 'ST57Q04',
 'ST57Q05',
 'MMINS',
 'LMINS',
 'SMINS',
 'PV1MATH',
 'PV2MATH',
 'PV3MATH',
 'PV4MATH',
 'PV5MATH',
 'PV1READ',
 'PV2READ',
 'PV3READ',
 'PV4READ',
 'PV5READ',
 'PV1SCIE',
 'PV2SCIE',
 'PV3SCIE',
 'PV4SCIE',
 'PV5SCIE']

### Create an average of all Math, Reading, and Science scores

For each of the Math, Reading, and Science scores, there are 5 scores provided. The variation between these scores is not needed, so rather we will look at the average for each of these major categories and remove the individual scores.

In [12]:
# Create average of Math scores
pisa_clean['Average Math Score']    = (pisa_clean['PV1MATH'] + pisa_clean['PV2MATH'] + pisa_clean['PV3MATH'] + pisa_clean['PV4MATH'] + pisa_clean['PV5MATH']) / 5

# Create average of Reading scores
pisa_clean['Average Reading Score'] = (pisa_clean['PV1READ'] + pisa_clean['PV2READ'] + pisa_clean['PV3READ'] + pisa_clean['PV4READ'] + pisa_clean['PV5READ']) / 5

# Create average of Science scores
pisa_clean['Average Science Score'] = (pisa_clean['PV1SCIE'] + pisa_clean['PV2SCIE'] + pisa_clean['PV3SCIE'] + pisa_clean['PV4SCIE'] + pisa_clean['PV5SCIE']) / 5

# Create average score
pisa_clean['Average Total Score']   = (pisa_clean['Average Math Score'] + pisa_clean['Average Reading Score'] + pisa_clean['Average Science Score'] +pisa_clean['Average Math Score']) / 3

In [13]:
# Remove the individual scores for Math, Reading, and Science
pisa_clean.drop(columns=['PV1MATH',  'PV2MATH', 'PV3MATH', 'PV4MATH', 'PV5MATH',   
                         'PV1READ', 'PV2READ',  'PV3READ', 'PV4READ',  'PV5READ',
                         'PV1SCIE',  'PV2SCIE', 'PV3SCIE',  'PV4SCIE', 'PV5SCIE'], inplace=True)

In [14]:
# Test to see accuracy of the Math score
print(pisa_clean['Average Math Score'].describe())
print('-'*22)

# Test to see accuracy of the Reading score
print(pisa_clean['Average Reading Score'].describe())
print('-'*22)

# Test to see accuracy of the Science score
print(pisa_clean['Average Science Score'].describe())

# Test to see accuracy of the Total score
print(pisa_clean['Average Total Score'].describe())

count    485490.000000
mean        469.651234
std         100.786610
min          54.767080
25%         396.019620
50%         465.734520
75%         540.123060
max         903.107960
Name: Average Math Score, dtype: float64
----------------------
count    485490.000000
mean        472.006964
std          98.863310
min           6.445400
25%         405.044200
50%         475.477980
75%         542.831195
max         849.359740
Name: Average Reading Score, dtype: float64
----------------------
count    485490.000000
mean        475.808094
std          97.998470
min          25.158540
25%         405.762800
50%         475.512860
75%         546.381920
max         857.832900
Name: Average Science Score, dtype: float64
count    485490.000000
mean        629.039175
std         128.839689
min         105.496473
25%         536.312167
50%         627.318087
75%         721.099990
max        1123.882273
Name: Average Total Score, dtype: float64


In [15]:
pisa_clean.sample(10)

STIDSTD ST04Q01                                ST13Q01 ST14Q01  \
195039     5923  Female                  <ISCED level 3B, 3C>      NaN   
458497     4121    Male                       <ISCED level 1>       No   
181051     5377    Male                      <ISCED level 3A>       No   
230733     3723    Male                      <ISCED level 3A>       No   
275340     3678    Male                      <ISCED level 3A>       No   
207611      711    Male                       <ISCED level 2>       No   
48967      7581    Male                      <ISCED level 3A>      NaN   
472632     2395    Male                       <ISCED level 2>      NaN   
113071     5829  Female                       <ISCED level 2>       No   
385612     2084    Male  She did not complete <ISCED level 1>       No   

       ST14Q02 ST14Q03 ST14Q04                               ST17Q01 ST18Q01  \
195039     NaN     NaN     NaN                 <ISCED level 3B, 3C>      NaN   
458497      No     NaN      No                      <ISCED level 1>       No   
181051     Yes      No      No                 <ISCED level 3B, 3C>       No   
230733      No      No      No                     <ISCED level 3A>       No   
275340     Yes      No     NaN                 <ISCED level 3B, 3C>       No   
207611      No      No     Yes                      <ISCED level 2>       No   
48967      NaN     Yes     NaN                     <ISCED level 3A>      NaN   
472632     NaN     NaN     NaN                     <ISCED level 3A>      Yes   
113071      No      No      No                     <ISCED level 3A>       No   
385612      No      No      No  He did not complete <ISCED level 1>       No   

       ST18Q02  ... ST57Q03 ST57Q04 ST57Q05  MMINS  LMINS  SMINS  \
195039     NaN  ...     NaN     NaN     NaN    NaN    NaN    NaN   
458497      No  ...     2.0     1.0     1.0   60.0  180.0   60.0   
181051      No  ...     0.0     0.0     2.0  135.0  135.0   90.0   
230733      No  ...     0.0     0.0     0.0  160.0  160.0  160.0   
275340      No  ...     0.0     0.0     2.0    NaN    NaN    NaN   
207611      No  ...     NaN     NaN     NaN    NaN    NaN    NaN   
48967      NaN  ...     0.0     0.0     1.0  200.0  150.0  100.0   
472632     Yes  ...     0.0     0.0     0.0  160.0  160.0  400.0   
113071      No  ...     0.0     0.0     1.0  135.0  135.0  270.0   
385612      No  ...     NaN     NaN     NaN    NaN    NaN    NaN   

        Average Math Score  Average Reading Score  Average Science Score  \
195039           459.50304              497.63924              487.26222   
458497           268.58546              320.27860              315.77770   
181051           415.72674              394.21796              479.52254   
230733           496.11312              484.83774              475.23310   
275340           358.55274              358.37100              397.83664   
207611           473.83550              502.48054              520.55202   
48967            601.58124              603.76620              614.82656   
472632           526.88112              474.73324              543.02498   
113071           507.56348              505.66176              503.67400   
385612           561.15440              599.67628              590.11562   

        Average Total Score  
195039           634.635847  
458497           391.075740  
181051           568.397993  
230733           650.765693  
275340           491.104373  
207611           656.901187  
48967            807.251747  
472632           690.506820  
113071           674.820907  
385612           770.700233  

[10 rows x 25 columns]

### Make the variable names readable

In the original dataframe, the variable names are all codes and unreadable without the dictionary. Therefore, we will change them to accurately reflect the information within the column.

In [16]:
# Rename columns
pisa_clean.rename({'STIDSTD': 'Student ID',
                   'ST04Q01': 'Gender', 
                   'ST13Q01': 'Mother - Highest Schooling', 
                   'ST14Q01': 'Mother Qualifications - ISCED level 6',
                   'ST14Q02': 'Mother Qualifications - ISCED level 5A',
                   'ST14Q03': 'Mother Qualifications - ISCED level 5B',
                   'ST14Q04': 'Mother Qualifications - ISCED level 4',
                   'ST17Q01': 'Father - Highest Schooling',
                   'ST18Q01': 'Father Qualifications - ISCED level 6',
                   'ST18Q02': 'Father Qualifications - ISCED level 5A',
                   'ST18Q03': 'Father Qualifications - ISCED level 5B',
                   'ST18Q04': 'Father Qualifications - ISCED level 4',
                   'ST28Q01': 'How many books at home', 
                   'ST57Q01': 'Out-of-School Study Time - Homework',
                   'ST57Q02': 'Out-of-School Study Time - Guided Homework',
                   'ST57Q03': 'Out-of-School Study Time - Personal Tutor',
                   'ST57Q04': 'Out-of-School Study Time - Commercial Company',
                   'ST57Q05': 'Out-of-School Study Time - With Parent',
                   'MMINS': 'Learning time (minutes per week)- <Mathematics>',
                   'LMINS': 'Learning time (minutes per week)  - <test language>',
                   'SMINS': 'Learning time (minutes per week) - <Science>'}, axis='columns', inplace=True)

In [17]:
# Check to see if the columns are correctly renamed
list(pisa_clean)

['Student ID',
 'Gender',
 'Mother - Highest Schooling',
 'Mother Qualifications - ISCED level 6',
 'Mother Qualifications - ISCED level 5A',
 'Mother Qualifications - ISCED level 5B',
 'Mother Qualifications - ISCED level 4',
 'Father - Highest Schooling',
 'Father Qualifications - ISCED level 6',
 'Father Qualifications - ISCED level 5A',
 'Father Qualifications - ISCED level 5B',
 'Father Qualifications - ISCED level 4',
 'How many books at home',
 'Out-of-School Study Time - Homework',
 'Out-of-School Study Time - Guided Homework',
 'Out-of-School Study Time - Personal Tutor',
 'Out-of-School Study Time - Commercial Company',
 'Out-of-School Study Time - With Parent',
 'Learning time (minutes per week)- <Mathematics>',
 'Learning time (minutes per week)  - <test language>',
 'Learning time (minutes per week) - <Science>',
 'Average Math Score',
 'Average Reading Score',
 'Average Science Score',
 'Average Total Score']

### Education
Multiple columns with the single variable. 

http://uis.unesco.org/sites/default/files/documents/international-standard-classification-of-education-isced-2011-en.pdf

In [18]:
father_edu = pisa_clean.copy()

In [19]:
father_edu[['<ISCED level 1>', 
            '<ISCED level 2>', 
            '<ISCED level 3A>', 
            '<ISCED level 3B, 3C>', 
            'He did not complete <ISCED level 1>']] = pd.get_dummies(father_edu['Father - Highest Schooling'])

In [20]:
father_edu.rename({'He did not complete <ISCED level 1>': '<ISCED level 0>',
                   '<ISCED level 3A>': '<ISCED level 3>', 
                   '<ISCED level 3B, 3C>': '<ISCED level 3>',
                   'Father Qualifications - ISCED level 4': '<ISCED level 4>',
                   'Father Qualifications - ISCED level 5A': '<ISCED level 5>',
                   'Father Qualifications - ISCED level 5B': '<ISCED level 5>',
                   'Father Qualifications - ISCED level 6': '<ISCED level 6>'}, axis='columns', inplace=True)

In [21]:
father_edu['<ISCED level 0>'] = father_edu['<ISCED level 0>'].astype(str)
father_edu['<ISCED level 1>'] = father_edu['<ISCED level 1>'].astype(str)
father_edu['<ISCED level 2>'] = father_edu['<ISCED level 2>'].astype(str)
father_edu['<ISCED level 3>'] = father_edu['<ISCED level 3>'].astype(str)

In [22]:
father_edu['<ISCED level 0>'] = father_edu['<ISCED level 0>'].replace(regex='1', value='<ISCED level 0>')
father_edu['<ISCED level 1>'] = father_edu['<ISCED level 1>'].replace(regex='1', value='<ISCED level 1>')
father_edu['<ISCED level 2>'] = father_edu['<ISCED level 2>'].replace(regex='1', value='<ISCED level 2>')
father_edu['<ISCED level 3>'] = father_edu['<ISCED level 3>'].replace(regex='1', value='<ISCED level 3>')
father_edu['<ISCED level 4>'] = father_edu['<ISCED level 4>'].replace(regex='Yes', value='<ISCED level 4>')
father_edu['<ISCED level 5>'] = father_edu['<ISCED level 5>'].replace(regex='Yes', value='<ISCED level 5>')
father_edu['<ISCED level 6>'] = father_edu['<ISCED level 6>'].replace(regex='Yes', value='<ISCED level 6>')

In [23]:
father_edu.sample(10)

Student ID  Gender             Mother - Highest Schooling  \
427616        4841  Female                      <ISCED level 3A>    
471236         999    Male                      <ISCED level 3A>    
205620        3845  Female                      <ISCED level 3A>    
166334       20752    Male                      <ISCED level 3A>    
179755        4081  Female                  <ISCED level 3B, 3C>    
466831        1442    Male  She did not complete <ISCED level 1>    
171555         660    Male                       <ISCED level 2>    
157866       12284    Male                      <ISCED level 3A>    
201066       11950    Male                      <ISCED level 3A>    
36034        13884  Female                      <ISCED level 3A>    

       Mother Qualifications - ISCED level 6  \
427616                                    No   
471236                                    No   
205620                                    No   
166334                                   Yes   
179755                                    No   
466831                                    No   
171555                                    No   
157866                                    No   
201066                                    No   
36034                                     No   

       Mother Qualifications - ISCED level 5A  \
427616                                     No   
471236                                     No   
205620                                    Yes   
166334                                    Yes   
179755                                     No   
466831                                     No   
171555                                     No   
157866                                     No   
201066                                     No   
36034                                      No   

       Mother Qualifications - ISCED level 5B  \
427616                                    Yes   
471236                                    Yes   
205620                                     No   
166334                                     No   
179755                                    Yes   
466831                                     No   
171555                                     No   
157866                                     No   
201066                                    Yes   
36034                                     Yes   

       Mother Qualifications - ISCED level 4  \
427616                                   Yes   
471236                                   NaN   
205620                                    No   
166334                                    No   
179755                                    No   
466831                                   NaN   
171555                                    No   
157866                                    No   
201066                                   Yes   
36034                                     No   

                  Father - Highest Schooling  <ISCED level 6>  \
427616                     <ISCED level 3A>                No   
471236                      <ISCED level 2>                No   
205620                     <ISCED level 3A>                No   
166334                     <ISCED level 3A>   <ISCED level 6>   
179755                 <ISCED level 3B, 3C>                No   
466831  He did not complete <ISCED level 1>                No   
171555                     <ISCED level 3A>                No   
157866                     <ISCED level 3A>                No   
201066                 <ISCED level 3B, 3C>                No   
36034                      <ISCED level 3A>                No   

        <ISCED level 5>  ... Learning time (minutes per week) - <Science>  \
427616  <ISCED level 5>  ...                                        450.0   
471236               No  ...                                        120.0   
205620  <ISCED level 5>  ...                                        200.0   
166334  <ISCED level 5>  ...                                          NaN   
179755   

In [24]:
father_edu = pd.melt(father_edu, 
                     id_vars=['Student ID', 'Gender',
                            'Mother - Highest Schooling', 
                            'Mother Qualifications - ISCED level 6', 
                            'Mother Qualifications - ISCED level 5A',
                            'Mother Qualifications - ISCED level 5B',
                            'Mother Qualifications - ISCED level 4',
                            'How many books at home',
                            'Out-of-School Study Time - Homework',
                            'Out-of-School Study Time - Guided Homework',
                            'Out-of-School Study Time - Personal Tutor',
                            'Out-of-School Study Time - Commercial Company',
                            'Out-of-School Study Time - With Parent',
                            'Learning time (minutes per week)- <Mathematics>',
                            'Learning time (minutes per week)  - <test language>',
                            'Learning time (minutes per week) - <Science>',
                            'Average Math Score',
                            'Average Reading Score',
                            'Average Science Score',
                            'Average Total Score'], var_name='Completed_edu - Father', value_name='Education - Father')


In [25]:
father_edu.sample(10)

Student ID  Gender             Mother - Highest Schooling  \
3655615       16596    Male                      <ISCED level 3A>    
3216559        2816  Female  She did not complete <ISCED level 1>    
1777768        5596  Female                       <ISCED level 2>    
2927271        9589  Female                      <ISCED level 3A>    
2069732        4601  Female                       <ISCED level 1>    
3189389        4787    Male                      <ISCED level 3A>    
2332980        1770    Male                  <ISCED level 3B, 3C>    
2049058       11085    Male                  <ISCED level 3B, 3C>    
1997106        5163  Female                  <ISCED level 3B, 3C>    
2102389       14847    Male                      <ISCED level 3A>    

        Mother Qualifications - ISCED level 6  \
3655615                                    No   
3216559                                    No   
1777768                                    No   
2927271                                    No   
2069732                                   NaN   
3189389                                   NaN   
2332980                                   Yes   
2049058                                    No   
1997106                                    No   
2102389                                   Yes   

        Mother Qualifications - ISCED level 5A  \
3655615                                     No   
3216559                                     No   
1777768                                     No   
2927271                                    Yes   
2069732                                    NaN   
3189389                                    Yes   
2332980                                    NaN   
2049058                                     No   
1997106                                     No   
2102389                                    Yes   

        Mother Qualifications - ISCED level 5B  \
3655615                                     No   
3216559                                     No   
1777768                                     No   
2927271                                     No   
2069732                                    NaN   
3189389                                    NaN   
2332980                                    NaN   
2049058                                     No   
1997106                                     No   
2102389                                    Yes   

        Mother Qualifications - ISCED level 4 How many books at home  \
3655615                                    No          26-100 books    
3216559                                    No           11-25 books    
1777768                                   NaN            0-10 books    
2927271                                    No           11-25 books    
2069732                                   NaN            0-10 books    
3189389                                   NaN            0-10 books    
2332980                                   NaN            0-10 books    
2049058                                    No         201-500 books    
1997106                                    No           11-25 books    
2102389                                   Yes         201-500 books    

         Out-of-School Study Time - Homework  \
3655615                                  NaN   
3216559                                 16.0   
1777768                                  5.0   
2927271                                  NaN   
2069732                                  1.0   
3189389                                  NaN   
2332980                                  NaN   
2049058                                  NaN   
1997106                                  1.0   
2102389                                  3.0   

         Out-of-School Study Time - Guided Homework  ...  \
3655615                                         NaN  ...   
3216559                                         4.0  ...   
1777768                                         NaN  ...   
2927271                           

In [26]:
father_edu = father_edu[(father_edu['Education - Father'] != '0')] 
father_edu = father_edu[(father_edu['Education - Father'] != 'No')]

In [27]:
father_edu.sample(70)

Student ID  Gender             Mother - Highest Schooling  \
401080          864  Female                      <ISCED level 3A>    
1827081        2016  Female                      <ISCED level 3A>    
3730010       15878    Male                       <ISCED level 1>    
227056           46    Male                      <ISCED level 3A>    
98437          2424  Female                      <ISCED level 3A>    
579718        19759    Male                      <ISCED level 3A>    
307165         1104    Male                      <ISCED level 3A>    
3687193        3712  Female                  <ISCED level 3B, 3C>    
2205726       23177  Female                  <ISCED level 3B, 3C>    
360369          920    Male                      <ISCED level 3A>    
3444573        4757    Male                       <ISCED level 2>    
220777         4199  Female                       <ISCED level 2>    
3680399        3269    Male                      <ISCED level 3A>    
3800868        2222    Male                      <ISCED level 3A>    
4233436           8    Male                  <ISCED level 3B, 3C>    
1890197         722  Female                      <ISCED level 3A>    
1361966        1736  Female                                    NaN   
1505032        7176    Male                      <ISCED level 3A>    
471572         1335  Female                       <ISCED level 1>    
1570338        6626    Male                      <ISCED level 3A>    
2365009         274  Female                      <ISCED level 3A>    
3443472        3656    Male                      <ISCED level 3A>    
3713396        4599  Female                      <ISCED level 3A>    
2915005        2066  Female                      <ISCED level 3A>    
2996430        9021  Female                      <ISCED level 3A>    
3463802       10107  Female                       <ISCED level 2>    
125371         2200  Female                                    NaN   
2146949        3214    Male                      <ISCED level 3A>    
839910          168    Male                       <ISCED level 2>    
316213          511  Female                                    NaN   
...             ...     ...                                    ...   
1108660        4580  Female                       <ISCED level 2>    
1684663        1183  Female                                    NaN   
18473          2231    Male                                    NaN   
3385236        2059    Male                       <ISCED level 2>    
2084485        4424    Male                      <ISCED level 3A>    
3385376        2199    Male                      <ISCED level 3A>    
470008         4619    Male                      <ISCED level 3A>    
96077            64  Female                  <ISCED level 3B, 3C>    
2226985        6325  Female                      <ISCED level 3A>    
2184155        1606  Female                      <ISCED level 3A>    
482103         1573    Male                       <ISCED level 2>    
386894         3366  Female                       <ISCED level 2>    
4303912        2448    Male                  <ISCED level 3B, 3C>    
456963         2587    Male                       <ISCED level 1>    
4745865        3569    Male                       <ISCED level 1>    
4108327        3019    Male                      <ISCED level 3A>    
3492373       19474  Female                      <ISCED level 3A>    
4080931        7895    Male                      <ISCED level 3A>    
1330805         376    Male                      <ISCED level 3A>    
3635926        1962  Female                      <ISCED level 3A>    
388568         5040    Male  She did not complete <ISCED level 1>    
3768294        1269    Male                  <ISCED level 3B, 3C>    
192319         3203  Female                  <ISCED level 3B, 3C>    
441582         3899    Male                  <ISCED level 3B, 3C>    
235209         3183  Female                       <ISCED level 2>    
3450232        1819

In [28]:
father_edu = father_edu[(father_edu['Completed_edu - Father'] != 'Father - Highest Schooling')] 

In [29]:
father_edu['Education - Father'].value_counts()

<ISCED level 3>    306459
<ISCED level 5>    185130
<ISCED level 2>     84329
<ISCED level 4>     77265
<ISCED level 1>     35938
<ISCED level 6>     29589
<ISCED level 0>     16535
Name: Education - Father, dtype: int64

In [30]:
father_edu.drop(columns=['Completed_edu - Father'], axis=1, inplace=True)

In [31]:
father_edu.sample(10)

Student ID  Gender Mother - Highest Schooling  \
1852743        7023    Male          <ISCED level 3A>    
3587545        4612    Male          <ISCED level 3A>    
1895253        1100  Female          <ISCED level 3A>    
3441107        1291    Male      <ISCED level 3B, 3C>    
1130979       14417  Female      <ISCED level 3B, 3C>    
2003616        6391  Female           <ISCED level 2>    
720395         2879  Female                        NaN   
3537948        1417    Male          <ISCED level 3A>    
2118592         958  Female      <ISCED level 3B, 3C>    
2578656        5624  Female           <ISCED level 1>    

        Mother Qualifications - ISCED level 6  \
1852743                                    No   
3587545                                   NaN   
1895253                                    No   
3441107                                    No   
1130979                                   NaN   
2003616                                    No   
720395                                    NaN   
3537948                                    No   
2118592                                   NaN   
2578656                                    No   

        Mother Qualifications - ISCED level 5A  \
1852743                                    Yes   
3587545                                    Yes   
1895253                                     No   
3441107                                     No   
1130979                                    Yes   
2003616                                     No   
720395                                     NaN   
3537948                                     No   
2118592                                    NaN   
2578656                                     No   

        Mother Qualifications - ISCED level 5B  \
1852743                                    Yes   
3587545                                    NaN   
1895253                                     No   
3441107                                     No   
1130979                                    NaN   
2003616                                     No   
720395                                     NaN   
3537948                                    Yes   
2118592                                    NaN   
2578656                                     No   

        Mother Qualifications - ISCED level 4 How many books at home  \
1852743                                   Yes         101-200 books    
3587545                                   NaN    More than 500 books   
1895253                                    No            0-10 books    
3441107                                    No          26-100 books    
1130979                                   NaN          26-100 books    
2003616                                    No           11-25 books    
720395                                    NaN                    NaN   
3537948                                    No          26-100 books    
2118592                                   NaN           11-25 books    
2578656                                    No          26-100 books    

         Out-of-School Study Time - Homework  \
1852743                                  NaN   
3587545                                  6.0   
1895253                                  3.0   
3441107                                  1.0   
1130979                                 12.0   
2003616                                  NaN   
720395                                   NaN   
3537948                                  1.0   
2118592                                  6.0   
2578656                                  5.0   

         Out-of-School Study Time - Guided Homework  ...  \
1852743                                         NaN  ...   
3587545                                         1.0  ...   
1895253                                         3.0  ...   
3441107                                         0.0  ...   
1130979                                        10.0  ...   
2003616                                       

In [32]:
father_edu.duplicated().sum()

327461

In [33]:
# https://stackoverflow.com/questions/33042777/removing-duplicates-from-pandas-dataframe-with-condition-for-retaining-original
father_edu['Education - Father'] = father_edu['Education - Father'].astype('category')

father_edu['Education - Father'] = father_edu['Education - Father'].cat.set_categories(['<ISCED level 0>', 
                                                                                        '<ISCED level 1>', 
                                                                                        '<ISCED level 2>', 
                                                                                        '<ISCED level 3>',
                                                                                        '<ISCED level 4>', 
                                                                                        '<ISCED level 5>', 
                                                                                        '<ISCED level 6>'], ordered=True)
father_edu.sort_values(['Education - Father'], inplace=True)
# father_edu_clean = father_edu.drop_duplicates('Education - Father')

In [34]:
father_edu.duplicated().sum()

327461

In [35]:
father_edu.shape

(1263067, 21)

### Mother Edu

In [104]:
mother_edu = father_edu.copy()

In [105]:
mother_edu.head()

Student ID  Gender Mother - Highest Schooling  \
4854888        4948  Female           <ISCED level 1>    
4854895        4955  Female           <ISCED level 2>    
4427204        2529    Male           <ISCED level 1>    
4490059        6551  Female           <ISCED level 1>    
4490057        6549  Female          <ISCED level 3A>    

        Mother Qualifications - ISCED level 6  \
4854888                                    No   
4854895                                    No   
4427204                                    No   
4490059                                    No   
4490057                                   NaN   

        Mother Qualifications - ISCED level 5A  \
4854888                                     No   
4854895                                     No   
4427204                                     No   
4490059                                     No   
4490057                                    Yes   

        Mother Qualifications - ISCED level 5B  \
4854888                                    NaN   
4854895                                    NaN   
4427204                                     No   
4490059                                     No   
4490057                                    NaN   

        Mother Qualifications - ISCED level 4 How many books at home  \
4854888                                    No         101-200 books    
4854895                                    No          26-100 books    
4427204                                    No            0-10 books    
4490059                                   NaN            0-10 books    
4490057                                   NaN            0-10 books    

         Out-of-School Study Time - Homework  \
4854888                                  2.0   
4854895                                  5.0   
4427204                                  2.0   
4490059                                  NaN   
4490057                                  4.0   

         Out-of-School Study Time - Guided Homework  ...  \
4854888                                         2.0  ...   
4854895                                         NaN  ...   
4427204                                         2.0  ...   
4490059                                         NaN  ...   
4490057                                         NaN  ...   

         Out-of-School Study Time - Commercial Company  \
4854888                                            0.0   
4854895                                            0.0   
4427204                                            3.0   
4490059                                            NaN   
4490057                                            3.0   

         Out-of-School Study Time - With Parent  \
4854888                                     0.0   
4854895                                     0.0   
4427204                                     1.0   
4490059                                     NaN   
4490057                                     4.0   

         Learning time (minutes per week)- <Mathematics>  \
4854888                                            315.0   
4854895                                              NaN   
4427204                                            135.0   
4490059                                              NaN   
4490057                                            360.0   

         Learning time (minutes per week)  - <test language>  \
4854888                                                NaN     
4854895                                                NaN     
4427204                                              180.0     
4490059                                                NaN     
4490057                                              360.0     

         Learning time (minutes per week) - <Science>  Average Math Score  \
4854888                                         450.0           441.04222   
4854895                                           NaN           486.22058   
4427204                                        

In [106]:
mother_edu['Mother - Highest Schooling'].value_counts()

<ISCED level 3A>                         661214
<ISCED level 3B, 3C>                     210837
<ISCED level 2>                          184488
<ISCED level 1>                           78980
She did not complete <ISCED level 1>      40916
Name: Mother - Highest Schooling, dtype: int64

In [107]:
pisa_clean['Mother - Highest Schooling'].value_counts()

<ISCED level 3A>                         236993
<ISCED level 3B, 3C>                      83048
<ISCED level 2>                           82614
<ISCED level 1>                           36556
She did not complete <ISCED level 1>      18768
Name: Mother - Highest Schooling, dtype: int64

In [108]:
mother_edu[['<ISCED level 1>', 
            '<ISCED level 2>', 
            '<ISCED level 3A>', 
            '<ISCED level 3B, 3C>', 
            'She did not complete <ISCED level 1>']] = pd.get_dummies(mother_edu['Mother - Highest Schooling'])

In [109]:
mother_edu.rename({'She did not complete <ISCED level 1>': '<ISCED level 0>',
                   '<ISCED level 3A>': '<ISCED level 3>', 
                   '<ISCED level 3B, 3C>': '<ISCED level 3>',
                   'Mother Qualifications - ISCED level 4': '<ISCED level 4>',
                   'Mother Qualifications - ISCED level 5A': '<ISCED level 5>',
                   'Mother Qualifications - ISCED level 5B': '<ISCED level 5>',
                   'Mother Qualifications - ISCED level 6': '<ISCED level 6>'}, axis='columns', inplace=True)

In [110]:
mother_edu['<ISCED level 0>'] = mother_edu['<ISCED level 0>'].astype(str)
mother_edu['<ISCED level 1>'] = mother_edu['<ISCED level 1>'].astype(str)
mother_edu['<ISCED level 2>'] = mother_edu['<ISCED level 2>'].astype(str)
mother_edu['<ISCED level 3>'] = mother_edu['<ISCED level 3>'].astype(str)

In [111]:
mother_edu['<ISCED level 0>'] = mother_edu['<ISCED level 0>'].replace(regex='1', value='<ISCED level 0>')
mother_edu['<ISCED level 1>'] = mother_edu['<ISCED level 1>'].replace(regex='1', value='<ISCED level 1>')
mother_edu['<ISCED level 2>'] = mother_edu['<ISCED level 2>'].replace(regex='1', value='<ISCED level 2>')
mother_edu['<ISCED level 3>'] = mother_edu['<ISCED level 3>'].replace(regex='1', value='<ISCED level 3>')
mother_edu['<ISCED level 4>'] = mother_edu['<ISCED level 4>'].replace(regex='Yes', value='<ISCED level 4>')
mother_edu['<ISCED level 5>'] = mother_edu['<ISCED level 5>'].replace(regex='Yes', value='<ISCED level 5>')
mother_edu['<ISCED level 6>'] = mother_edu['<ISCED level 6>'].replace(regex='Yes', value='<ISCED level 6>')

In [112]:
mother_edu.sample(10)

Student ID  Gender Mother - Highest Schooling <ISCED level 6>  \
2389262        3708    Male           <ISCED level 2>               No   
2243974        1211  Female          <ISCED level 3A>               No   
1608272        6220    Male          <ISCED level 3A>               No   
3694684          69    Male          <ISCED level 3A>               No   
2279691       22029    Male          <ISCED level 3A>               No   
1051597        6148    Male          <ISCED level 3A>               No   
3404577        1405  Female          <ISCED level 3A>               No   
1555886        3403  Female      <ISCED level 3B, 3C>              NaN   
3443836        4020    Male                        NaN             NaN   
2217820        4198  Female          <ISCED level 3A>               No   

         <ISCED level 5>  <ISCED level 5>  <ISCED level 4>  \
2389262               No               No              NaN   
2243974               No  <ISCED level 5>               No   
1608272  <ISCED level 5>               No               No   
3694684               No               No  <ISCED level 4>   
2279691               No               No              NaN   
1051597  <ISCED level 5>               No               No   
3404577               No               No               No   
1555886              NaN              NaN  <ISCED level 4>   
3443836              NaN              NaN              NaN   
2217820               No               No              NaN   

        How many books at home  Out-of-School Study Time - Homework  \
2389262          26-100 books                                   2.0   
2243974         101-200 books                                   NaN   
1608272    More than 500 books                                  NaN   
3694684           11-25 books                                   NaN   
2279691          26-100 books                                   4.0   
1051597         201-500 books                                   2.0   
3404577          26-100 books                                   7.0   
1555886         201-500 books                                   1.0   
3443836         101-200 books                                   6.0   
2217820         101-200 books                                  14.0   

         Out-of-School Study Time - Guided Homework  ...  Average Math Score  \
2389262                                         0.0  ...           502.26670   
2243974                                         NaN  ...           463.00824   
1608272                                         NaN  ...           459.81462   
3694684                                         NaN  ...           506.08350   
2279691                                         1.0  ...           526.18008   
1051597                                         0.0  ...           604.93068   
3404577                                         NaN  ...           388.54184   
1555886                                         1.0  ...           522.59698   
3443836                                         0.0  ...           407.85948   
2217820                                         7.0  ...           433.87600   

         Average Reading Score  Average Science Score  Average Total Score  \
2389262              547.71024              524.84148           692.361707   
2243974              544.42410              469.91794           646.786173   
1608272              422.60680              487.16896           609.801667   
3694684              561.74428              567.64268           713.851320   
2279691              498.39060              539.48152           696.744093   
1051597              611.86584              614.36032           812.029173   
3404577              412.88636              378.90714           522.959060   
1555886              543.55036              552.72286           713.822393   
3443836              367.19240              426.65052           536.520627   
2217820              459.90952              481.66726           603.

In [113]:
mother_edu = pd.melt(mother_edu, 
                     id_vars=['Student ID', 'Gender',
                            'How many books at home',
                            'Out-of-School Study Time - Homework',
                            'Out-of-School Study Time - Guided Homework',
                            'Out-of-School Study Time - Personal Tutor',
                            'Out-of-School Study Time - Commercial Company',
                            'Out-of-School Study Time - With Parent',
                            'Learning time (minutes per week)- <Mathematics>',
                            'Learning time (minutes per week)  - <test language>',
                            'Learning time (minutes per week) - <Science>',
                            'Average Math Score',
                            'Average Reading Score',
                            'Average Science Score',
                            'Average Total Score',
                             'Education - Father'], var_name='Completed_edu - Mother', value_name='Education - Mother')


In [114]:
mother_edu.sample(10)

Student ID  Gender How many books at home  \
6952878         4794  Female          26-100 books    
7520136        12443    Male         101-200 books    
10906027        4590    Male           11-25 books    
11933010        1289  Female         201-500 books    
2190198         7599  Female         101-200 books    
3575001          570  Female         101-200 books    
8571029         2328    Male          26-100 books    
137298          2864  Female            0-10 books    
12397477        1287    Male            0-10 books    
8753362         2783    Male                    NaN   

          Out-of-School Study Time - Homework  \
6952878                                   3.0   
7520136                                   2.0   
10906027                                  NaN   
11933010                                 28.0   
2190198                                   5.0   
3575001                                   NaN   
8571029                                   3.0   
137298                                    NaN   
12397477                                  NaN   
8753362                                   NaN   

          Out-of-School Study Time - Guided Homework  \
6952878                                          2.0   
7520136                                          0.0   
10906027                                         NaN   
11933010                                         3.0   
2190198                                          0.0   
3575001                                          NaN   
8571029                                          2.0   
137298                                           NaN   
12397477                                         NaN   
8753362                                          NaN   

          Out-of-School Study Time - Personal Tutor  \
6952878                                         0.0   
7520136                                         2.0   
10906027                                        NaN   
11933010                                        0.0   
2190198                                         0.0   
3575001                                         NaN   
8571029                                         2.0   
137298                                          NaN   
12397477                                        NaN   
8753362                                         NaN   

          Out-of-School Study Time - Commercial Company  \
6952878                                             4.0   
7520136                                             0.0   
10906027                                            NaN   
11933010                                            0.0   
2190198                                             0.0   
3575001                                             NaN   
8571029                                             6.0   
137298                                              NaN   
12397477                                            NaN   
8753362                                             NaN   

          Out-of-School Study Time - With Parent  \
6952878                                      0.0   
7520136                                      0.0   
10906027                                     NaN   
11933010                                     2.0   
2190198                                      0.0   
3575001                                      NaN   
8571029                                      2.0   
137298                                       NaN   
12397477                                     NaN   
8753362                                      NaN   

          Learning time (minutes per week)- <Mathematics>  \
6952878                                             160.0   
7520136                                             200.0   
10906027                                              NaN   
11933010                                             90.0   
2190198                                             225.0   
3575001                                             

In [115]:
mother_edu = mother_edu[(mother_edu['Education - Mother'] != '0')] 
mother_edu = mother_edu[(mother_edu['Education - Mother'] != 'No')]

In [116]:
mother_edu.sample(70)

Student ID  Gender How many books at home  \
5732652         2095    Male            0-10 books    
683524          1993    Male         201-500 books    
4482314        30416  Female           11-25 books    
2832714         3391  Female         101-200 books    
6356183         3153  Female            0-10 books    
9587483         2656  Female         201-500 books    
3566333          711    Male         101-200 books    
6357327         7314  Female            0-10 books    
7652957        16042  Female          26-100 books    
5601685          651    Male            0-10 books    
4915369         5609  Female          26-100 books    
737464          1491  Female            0-10 books    
9079744         5409  Female            0-10 books    
1176751          424  Female           11-25 books    
7406059        10455  Female            0-10 books    
8623415        10342    Male         101-200 books    
3544416         4524  Female                    NaN   
285140          3539  Female          26-100 books    
8892114        19180    Male            0-10 books    
5596720         4929    Male           11-25 books    
10418922        9918    Male            0-10 books    
5253379        12069    Male         201-500 books    
11605186        6134  Female            0-10 books    
6194719          261  Female          26-100 books    
331663          1143    Male    More than 500 books   
6059182         4052  Female            0-10 books    
7112774         2387    Male            0-10 books    
6239161         1217  Female           11-25 books    
2789249         6051  Female    More than 500 books   
10420303        8595  Female            0-10 books    
...              ...     ...                    ...   
9578609          784    Male           11-25 books    
1714444         3582    Male          26-100 books    
6035242         1668  Female         201-500 books    
1099815         2339    Male           11-25 books    
1179546        25637    Male    More than 500 books   
446872         10470    Male          26-100 books    
3298761        20968  Female            0-10 books    
9063150         8430  Female    More than 500 books   
3386344        13269    Male           11-25 books    
5586395          363  Female         101-200 books    
3667012         7047    Male            0-10 books    
4983619         2273  Female         101-200 books    
5853596         3410  Female            0-10 books    
7716921         4447    Male            0-10 books    
2067768          125    Male          26-100 books    
1369316         1187    Male         101-200 books    
2039977         7469    Male         201-500 books    
9118222         8727    Male            0-10 books    
742320           620  Female           11-25 books    
8866315        15000    Male    More than 500 books   
2473774         1755  Female            0-10 books    
6106168          743    Male         201-500 books    
4597329          948  Female                    NaN   
9342160         4509    Male           11-25 books    
778070          4417    Male          26-100 books    
3179073         2435  Female           11-25 books    
9805168         4823    Male            0-10 books    
480811          4905  Female          26-100 books    
4761239         3113    Male          26-100 books    
5963663         5525    Male          26-100 books    

          Out-of-School Study Time - Homework  \
5732652                                   NaN   
683524                                   10.0   
4482314                                   NaN   
2832714                                   1.0   
6356183                                   NaN   
9587483                                   6.0   
3566333                                   5.0   
6357327                                   NaN   
7652957                                  12.0   
5601685                                   2.0   
4915369                                   4.0   
737464     

In [117]:
mother_edu = mother_edu[(mother_edu['Completed_edu - Mother'] != 'Mother - Highest Schooling')] 

In [118]:
mother_edu['Education - Mother'].value_counts()

<ISCED level 3>    872051
<ISCED level 5>    596385
<ISCED level 4>    206129
<ISCED level 2>    184488
<ISCED level 6>     85227
<ISCED level 1>     78980
<ISCED level 0>     40916
Name: Education - Mother, dtype: int64

In [119]:
mother_edu.drop(columns=['Completed_edu - Mother'], axis=1, inplace=True)

In [120]:
mother_edu.sample(10)

Student ID  Gender How many books at home  \
5845762       10364    Male            0-10 books    
3087139        1947    Male          26-100 books    
3349217         478  Female            0-10 books    
5546856        2220    Male         201-500 books    
3780455        1363  Female         101-200 books    
6131037        4501  Female                    NaN   
5525469        3542  Female         101-200 books    
4055437        2762  Female           11-25 books    
4033695        7965  Female         101-200 books    
2382615        1577  Female            0-10 books    

         Out-of-School Study Time - Homework  \
5845762                                  NaN   
3087139                                 10.0   
3349217                                  3.0   
5546856                                 12.0   
3780455                                  7.0   
6131037                                  NaN   
5525469                                  NaN   
4055437                                  3.0   
4033695                                  3.0   
2382615                                  2.0   

         Out-of-School Study Time - Guided Homework  \
5845762                                         NaN   
3087139                                         5.0   
3349217                                         0.0   
5546856                                         0.0   
3780455                                         4.0   
6131037                                         NaN   
5525469                                         NaN   
4055437                                         0.0   
4033695                                         NaN   
2382615                                         0.0   

         Out-of-School Study Time - Personal Tutor  \
5845762                                        NaN   
3087139                                        NaN   
3349217                                        0.0   
5546856                                        2.0   
3780455                                        0.0   
6131037                                        NaN   
5525469                                        NaN   
4055437                                        0.0   
4033695                                        0.0   
2382615                                        4.0   

         Out-of-School Study Time - Commercial Company  \
5845762                                            NaN   
3087139                                            NaN   
3349217                                            0.0   
5546856                                            2.0   
3780455                                            0.0   
6131037                                            NaN   
5525469                                            NaN   
4055437                                            0.0   
4033695                                            0.0   
2382615                                            0.0   

         Out-of-School Study Time - With Parent  \
5845762                                     NaN   
3087139                                     5.0   
3349217                                     0.0   
5546856                                     0.0   
3780455                                     1.0   
6131037                                     NaN   
5525469                                     NaN   
4055437                                     0.0   
4033695                                     3.0   
2382615                                     0.0   

         Learning time (minutes per week)- <Mathematics>  \
5845762                                              NaN   
3087139                                            135.0   
3349217                                              NaN   
5546856                                            220.0   
3780455                                            240.0   
6131037                                              NaN   
5525469                                              NaN   
405

In [121]:
mother_edu['Student ID'].duplicated()

1263071     False
1263074     False
1263083     False
1263085     False
1263090     False
1263091     False
1263093     False
1263095     False
1263111     False
1263112     False
1263121     False
1263141     False
1263144     False
1263146     False
1263161     False
1263166     False
1263168     False
1263172     False
1263182     False
1263187     False
1263191     False
1263196     False
1263201     False
1263203     False
1263208     False
1263221     False
1263225     False
1263228     False
1263229     False
1263237     False
            ...  
12628706     True
12628886     True
12629156     True
12629267     True
12629304     True
12629305     True
12629377     True
12629499     True
12629520     True
12629610     True
12629647     True
12629674     True
12629695     True
12629723     True
12629728     True
12629782     True
12629808     True
12629854     True
12629922     True
12629936     True
12629960     True
12629966     True
12630019     True
12630067     True
12630142  

In [122]:
mother_edu.replace(["NaN"], np.nan, inplace = True)
mother_edu.dropna(inplace = True)

In [123]:
mother_edu['Education - Mother'] = mother_edu['Education - Mother'].astype('category')

mother_edu['Education - Mother'] = mother_edu['Education - Mother'].cat.set_categories(['<ISCED level 0>', 
                                                                                        '<ISCED level 1>', 
                                                                                        '<ISCED level 2>', 
                                                                                        '<ISCED level 3>',
                                                                                        '<ISCED level 4>', 
                                                                                        '<ISCED level 5>', 
                                                                                        '<ISCED level 6>'], ordered=True)
mother_edu.sort_values(['Education - Mother'], inplace=True)
mother_edu_cleaner = mother_edu.drop_duplicates(subset='Student ID', keep='last')

In [124]:
mother_edu['Student ID'].duplicated().sum()

575333

In [125]:
mother_edu.shape

(603927, 17)

In [126]:
mother_edu_cleaner.shape

(28594, 17)

In [127]:
mother_edu_cleaner.duplicated().sum()

0

In [128]:
mother_edu_cleaner.sample(10)

Student ID  Gender How many books at home  \
2800468        17548    Male          26-100 books    
5565876        22560    Male          26-100 books    
4359619        21001  Female         201-500 books    
8967933        16781    Male          26-100 books    
9539786        24044  Female            0-10 books    
2699452         8846  Female         201-500 books    
1984811         4091  Female         201-500 books    
11371588       15593    Male            0-10 books    
7879269        26246  Female         101-200 books    
2684640         7000  Female          26-100 books    

          Out-of-School Study Time - Homework  \
2800468                                   1.0   
5565876                                   2.0   
4359619                                   6.0   
8967933                                  10.0   
9539786                                   8.0   
2699452                                   9.0   
1984811                                  24.0   
11371588                                  1.0   
7879269                                   4.0   
2684640                                   3.0   

          Out-of-School Study Time - Guided Homework  \
2800468                                          0.0   
5565876                                          0.0   
4359619                                          5.0   
8967933                                          0.0   
9539786                                          0.0   
2699452                                          0.0   
1984811                                          3.0   
11371588                                         0.0   
7879269                                          0.0   
2684640                                          1.0   

          Out-of-School Study Time - Personal Tutor  \
2800468                                         0.0   
5565876                                         0.0   
4359619                                         2.0   
8967933                                         2.0   
9539786                                         0.0   
2699452                                         0.0   
1984811                                         3.0   
11371588                                        0.0   
7879269                                         0.0   
2684640                                         1.0   

          Out-of-School Study Time - Commercial Company  \
2800468                                             0.0   
5565876                                             0.0   
4359619                                             0.0   
8967933                                             4.0   
9539786                                             0.0   
2699452                                             0.0   
1984811                                             0.0   
11371588                                            0.0   
7879269                                             0.0   
2684640                                             0.0   

          Out-of-School Study Time - With Parent  \
2800468                                      0.0   
5565876                                      0.0   
4359619                                      5.0   
8967933                                      0.0   
9539786                                      0.0   
2699452                                      1.0   
1984811                                      2.0   
11371588                                     0.0   
7879269                                      0.0   
2684640                                      1.0   

          Learning time (minutes per week)- <Mathematics>  \
2800468                                             260.0   
5565876                                             200.0   
4359619                                             375.0   
8967933                                             180.0   
9539786                                             240.0   
2699452                                             

In [129]:
pisa_clean = mother_edu.copy()

In [130]:
pisa_clean.shape

(603927, 17)

In [131]:
pisa_clean = mother_edu.copy()

In [132]:
pisa_clean.shape

(603927, 17)

### 4. Missing values in relevant columns 
('Mother Highest Schooling', 'Father Highest Schooling', 'How many books at home', 'Out-of-School Study Time - Homework', 'Out-of-School Study Time - Guided Homework', 'Out-of-School Study Time - Personal Tutor','Out-of-School Study Time - Commercial Company', 'Out-of-School Study Time - With Parent', 'Learning time (minutes per week)- Mathematics', 'Learning time (minutes per week)  - test language', 'Learning time (minutes per week) - Science')

Due to the amount of rows in this dataframe, we can comfortably remove the rows with missing values.

In [ ]:

pisa_clean.replace(["NaN"], np.nan, inplace = True)
pisa_clean.dropna(inplace = True)

In [133]:
# Check to see if null values correctly dropped
pisa_clean

Student ID  Gender How many books at home  \
12102512        1218    Male           11-25 books    
11385412        6919    Male            0-10 books    
11385419        6994    Male            0-10 books    
11385428        6257  Female            0-10 books    
11385459        6136  Female            0-10 books    
11385486        6566  Female          26-100 books    
11385573        8033  Female          26-100 books    
11385582        8076  Female            0-10 books    
11385583        8082    Male           11-25 books    
11385609        7969  Female            0-10 books    
11385610        7975  Female          26-100 books    
11385640        7334  Female            0-10 books    
11385694        7523  Female            0-10 books    
11385710        7449  Female           11-25 books    
11385728        1309    Male          26-100 books    
11385747        1161  Female            0-10 books    
11385757        1214  Female            0-10 books    
11385841         813    Male           11-25 books    
11385369        6740    Male           11-25 books    
11385862         800    Male            0-10 books    
11385363        6701  Female            0-10 books    
11385339        6825    Male           11-25 books    
11384933        8786  Female           11-25 books    
11384966        8666  Female            0-10 books    
11384970        8706  Female            0-10 books    
11384991        9061    Male            0-10 books    
11385002        8965  Female            0-10 books    
11385008        8912    Male            0-10 books    
11385022        8962    Male            0-10 books    
11385050        8208    Male            0-10 books    
...              ...     ...                    ...   
1769825         3336  Female          26-100 books    
1769715          572  Female         201-500 books    
1769959         4270  Female          26-100 books    
1769675         5048  Female         101-200 books    
1769467         7356    Male          26-100 books    
1769415         2428    Male          26-100 books    
1769390         7672  Female         201-500 books    
1769348         4085    Male            0-10 books    
1769345         7627  Female         201-500 books    
1769238         7167    Male    More than 500 books   
1769537         4608    Male         101-200 books    
1770950          598  Female           11-25 books    
1769963         4277    Male            0-10 books    
1770071         5057  Female         201-500 books    
1770887         1518    Male         201-500 books    
1770861         4610  Female            0-10 books    
1770663        18698  Female         101-200 books    
1770643        18713  Female          26-100 books    
1770539         5320  Female         101-200 books    
1770440         4321    Male    More than 500 books   
1769969         4231    Male          26-100 books    
1770420         5307    Male         201-500 books    
1770394          478  Female          26-100 books    
1770369         5302    Male          26-100 books    
1770286         1338  Female         201-500 books    
1770222         1862  Female    More than 500 books   
1770218         5271    Male           11-25 books    
1770077         5056    Male         201-500 books    
1770408         4498  Female           11-25 books    
1263271         7089  Female           11-25 books    

          Out-of-School Study Time - Homework  \
12102512                                  2.0   
11385412                                  2.0   
11385419                                  6.0   
11385428                                 10.0   
11385459                                  7.0   
11385486                                 12.0   
11385573                                  2.0   
11385582                                  4.0   
11385583                                  1.0   
11385609                                 10.0   
11385610                                 14.0   
11385640   

In [134]:
# Confirm with smaller shape of dataframe
pisa_clean.shape

(603927, 17)

### 5. Create a total for Out-of-School Study Time and Learning time (minutes per week)



In [135]:
# Create total of Out-of-School Study Time
pisa_clean['Out-of-School Study Time - Total'] = pisa_clean['Out-of-School Study Time - Homework'] + pisa_clean['Out-of-School Study Time - Guided Homework'] + pisa_clean['Out-of-School Study Time - Personal Tutor'] + pisa_clean['Out-of-School Study Time - Commercial Company'] + pisa_clean['Out-of-School Study Time - With Parent']

# Create total of Learning time (minutes per week)
pisa_clean['Learning time (minutes per week) - Total'] = pisa_clean['Learning time (minutes per week)- <Mathematics>'] + pisa_clean['Learning time (minutes per week)  - <test language>'] + pisa_clean['Learning time (minutes per week) - <Science>'] 
                                                                                                                 

In [136]:
list(pisa_clean)

['Student ID',
 'Gender',
 'How many books at home',
 'Out-of-School Study Time - Homework',
 'Out-of-School Study Time - Guided Homework',
 'Out-of-School Study Time - Personal Tutor',
 'Out-of-School Study Time - Commercial Company',
 'Out-of-School Study Time - With Parent',
 'Learning time (minutes per week)- <Mathematics>',
 'Learning time (minutes per week)  - <test language>',
 'Learning time (minutes per week) - <Science>',
 'Average Math Score',
 'Average Reading Score',
 'Average Science Score',
 'Average Total Score',
 'Education - Father',
 'Education - Mother',
 'Out-of-School Study Time - Total',
 'Learning time (minutes per week) - Total']

<a id='assessing2'></a>
### Final Assessment

In this section, we will do the preliminary visual and programmatic assessment of the pisa datasets to determine whether or not it holds any major quality or tidiness issues.

In [137]:
pisa_clean

Student ID  Gender How many books at home  \
12102512        1218    Male           11-25 books    
11385412        6919    Male            0-10 books    
11385419        6994    Male            0-10 books    
11385428        6257  Female            0-10 books    
11385459        6136  Female            0-10 books    
11385486        6566  Female          26-100 books    
11385573        8033  Female          26-100 books    
11385582        8076  Female            0-10 books    
11385583        8082    Male           11-25 books    
11385609        7969  Female            0-10 books    
11385610        7975  Female          26-100 books    
11385640        7334  Female            0-10 books    
11385694        7523  Female            0-10 books    
11385710        7449  Female           11-25 books    
11385728        1309    Male          26-100 books    
11385747        1161  Female            0-10 books    
11385757        1214  Female            0-10 books    
11385841         813    Male           11-25 books    
11385369        6740    Male           11-25 books    
11385862         800    Male            0-10 books    
11385363        6701  Female            0-10 books    
11385339        6825    Male           11-25 books    
11384933        8786  Female           11-25 books    
11384966        8666  Female            0-10 books    
11384970        8706  Female            0-10 books    
11384991        9061    Male            0-10 books    
11385002        8965  Female            0-10 books    
11385008        8912    Male            0-10 books    
11385022        8962    Male            0-10 books    
11385050        8208    Male            0-10 books    
...              ...     ...                    ...   
1769825         3336  Female          26-100 books    
1769715          572  Female         201-500 books    
1769959         4270  Female          26-100 books    
1769675         5048  Female         101-200 books    
1769467         7356    Male          26-100 books    
1769415         2428    Male          26-100 books    
1769390         7672  Female         201-500 books    
1769348         4085    Male            0-10 books    
1769345         7627  Female         201-500 books    
1769238         7167    Male    More than 500 books   
1769537         4608    Male         101-200 books    
1770950          598  Female           11-25 books    
1769963         4277    Male            0-10 books    
1770071         5057  Female         201-500 books    
1770887         1518    Male         201-500 books    
1770861         4610  Female            0-10 books    
1770663        18698  Female         101-200 books    
1770643        18713  Female          26-100 books    
1770539         5320  Female         101-200 books    
1770440         4321    Male    More than 500 books   
1769969         4231    Male          26-100 books    
1770420         5307    Male         201-500 books    
1770394          478  Female          26-100 books    
1770369         5302    Male          26-100 books    
1770286         1338  Female         201-500 books    
1770222         1862  Female    More than 500 books   
1770218         5271    Male           11-25 books    
1770077         5056    Male         201-500 books    
1770408         4498  Female           11-25 books    
1263271         7089  Female           11-25 books    

          Out-of-School Study Time - Homework  \
12102512                                  2.0   
11385412                                  2.0   
11385419                                  6.0   
11385428                                 10.0   
11385459                                  7.0   
11385486                                 12.0   
11385573                                  2.0   
11385582                                  4.0   
11385583                                  1.0   
11385609                                 10.0   
11385610                                 14.0   
11385640   

In [138]:
pisa_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 603927 entries, 12102512 to 1263271
Data columns (total 19 columns):
Student ID                                             603927 non-null int64
Gender                                                 603927 non-null object
How many books at home                                 603927 non-null object
Out-of-School Study Time - Homework                    603927 non-null float64
Out-of-School Study Time - Guided Homework             603927 non-null float64
Out-of-School Study Time - Personal Tutor              603927 non-null float64
Out-of-School Study Time - Commercial Company          603927 non-null float64
Out-of-School Study Time - With Parent                 603927 non-null float64
Learning time (minutes per week)- <Mathematics>        603927 non-null float64
Learning time (minutes per week)  - <test language>    603927 non-null float64
Learning time (minutes per week) - <Science>           603927 non-null float64
Average Math Sco

<a id='sav'></a>
## Store

In [139]:
# Store the clean DataFrame in a CSV file
pisa_clean.to_csv("pisa_df.csv", index=False, encoding='utf8')

<a id='conclusion'></a>
## Conclusion

<a id='references'></a>
## References